src/utils/diff.rs

```diff
@@ -0,0 +1,16 @@
+/// A trait for computing the difference between two objects.
+pub trait DiffT<K: Ord + Clone, V: Clone> {
+    type DiffType;
+
+    /// Returns a `Self::DiffType` object that represents the difference between this object and
+    /// other.
+    fn diff(&self, other: &Self) -> Self::DiffType;
+}
```

Comment:
Should we just call these `Diff` and `ApplyDiff` (without the `T` at the end)? Does this cause any issues somewhere?